### PR TITLE
Feature/905/street map layout

### LINKED ISSUE #904: Integration of Streetlayout as an optional layout algorithm
# Feature request

## Description

Referring to issue #788 

The user should have the possibility to switch the layout between the standard Squarified TreeMap and a street layout, where file nodes are displayed as buildings and directories are displayed as streets. A street layout has the advantage of a more apparent directory structure and stable positioning of nodes after metric changes.

A layout algorithm only influences the position and shape of a node's base area. Therefore the following 2D representations show the concepts of both proposed layouts algorithms. 

Example Squarified Treemap

Squarified TreeMap representations use less space, but nodes overlap each other, resulting in a less obvious directory structure representation. With this approach nodes are layed out to be "square-like" e.g. have aspect ratios near 1. This comes at the cost of the nodes' initial order, which has to be altered for the algorithm to produce good results. By sorting nodes by their size, a node's position inside the TreeMap is rather unstable for new images after metric changes.

Example Street Layout
Street Layouts tend to produce results that are larger in size than TreeMap layouts. A leaf nodes base area can have any form, but the most obvious is a square. This is for city metaphor reasons and node comparability. By streets branching from each other, all linked to a main "root street", the projects directory structure can be followed through when looking for a specific node. This results in stable results for the nodes' positions. 

To compare 2D layouts side by side, [this](https://github.com/reitermaniacL/software-map-visualizer "Software Map Visualizer") tool can be used.

## Acceptance criteria
- layout can be switched to street layout

## Open questions
- Since treemaps use the given space more efficiently than the street layout, should there be an option to substitute streets for treemaps at a certain depth into the project tree to reduce the overall size of the layout?
Example Street Layout

This would result in a layout like this in CodeCharta-Visualization 03/2020 with a custom Squarified TreeMap implementation and TreeMaps for direcotries with a depth > 3:

Example TMStreet Layout

- Can the TreeMapGenerator be modified to create treemaps at a specified position? The Street Layout implementation can already create TreeMaps inside the Street structure. If we use the existing implementation, we can use the functionality already available to us. On the other hand, different TreeMap types and adjustments in the algorithm are possible with a custom implementation.
- The layouts have a huge difference in size. Does the size of the street layout have to be scaled up or does the treemap layout have to be scaled down? With the mapSize attribute inside the treeMap settings one can specify the size of the TreeMap. This is not easily possible for a Street Layout since the size of the resulting layout is dependent on the project's size and structure. The Street Layout currently uses raw metric values and applies no scaling to a node's size. Is scaling for the 
TreeMap required, if zooming in and out is a subjectively more intuitive approach to look at smaller structures?

### PR DESCRIPTION
**Pull request by Med-H**
# Integration of street layout & fixing e2e tests

closes #904  & #905 & #1322 

## Description

User is able to choose between 'Squarified Treemap' and 'Street Layout' in options which then redraws the map with the corresponding layout

streetLayout

## TODOS

- [ ] Refactoring of the street algorithm needs more work, maybe abstract the layout Generator for both street and squared layout.
- [ ] Add unit tests.
- [ ] Add street Labels
- [ ] Apply same scaling on the current and street layout
- [ ] Add manual margin
- [ ] Improve label height positioning, not by using an empirical offset value
- [ ] Improve auto camera centering
- [ ] Check and working through the todos

### CODE REVIEW

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetLayoutGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		const mergedMap = StreetViewHelper.mergeDirectories(map, metricName)
+		const maxTreeMapFiles = state.appSettings.maxTreeMapFiles
+		const childBoxes = this.createBoxes(mergedMap, metricName, state, StreetOrientation.Vertical, 0, maxTreeMapFiles)
+		const rootStreet = new HorizontalStreet(mergedMap, childBoxes, 0)
+		rootStreet.calculateDimension(metricName)
+		const margin = state.dynamicSettings.margin * MARGIN_SCALING_FACTOR
+		const layoutNodes: CodeMapNode[] = rootStreet.layout(new Vector2(0, 0), margin)
+		
+		return layoutNodes.map(streetLayoutNode => {
+			return StreetViewHelper.buildNodeFrom(streetLayoutNode as CodeMapNode, heightScale, maxHeight, state, isDeltaState)
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/streetLayout/streetLayoutGenerator.ts:31]: We might want to extend the streetLayoutNode from `CodeMapNode`. That way it should be fine to pass the node through without forcing it.

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	}
+	return totalSize
+}
+
+function mergeDirectories(node: CodeMapNode, metricName: string): CodeMapNode {
+	let mergedNode = node
+	for (const child of node.children) {
+		if (!isLeaf(child)) {
+			const nodeSize = calculateSize(node, metricName)
+			const childSize = calculateSize(child, metricName)
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts:22]: Please only calculate the `nodeSize` once.

```suggestion
	let mergedNode = node
	const nodeSize = calculateSize(node, metricName)
	if (nodeSize === 0) {
		return mergedNode
	}
	for (const child of node.children) {
		if (!isLeaf(child)) {
			const childSize = calculateSize(child, metricName)
			if (childSize === 0)
				continue
```

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+			if (nodeSize === childSize) {
+				const nodeName = mergedNode.name
+				mergedNode = child
+				mergedNode.name = `${nodeName}/${child.name}`
+				break
+			}
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts:28]:
```suggestion
			}
			break
```

This is in combination with the comment above. If any child is not of size zero but also not the total nodeSize, it's impossible to have any other child that would fit the requirement. At least if I understood the logic correct.

File: visualization/app/codeCharta/util/algorithm/streetLayout/strip/horizontalStrip.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	worstAspectRatio(nodes: CodeMapNode[], rect: Rectangle, rootSize: number, metricName: string): number {
+		const width = rect.width
+		const rootArea = rect.area()
+		const totalSize = this.totalScaledSize(nodes, metricName, rootSize, rootArea)
+		const stripMin = this.min(nodes, metricName, rootSize, rootArea)
+		const stripMax = this.max(nodes, metricName, rootSize, rootArea)
+		const widthSquared = Math.pow(width, 2)
+		const totalSizeSquared = Math.pow(totalSize, 2)
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/streetLayout/strip/horizontalStrip.ts:54]:
```suggestion
		const widthSquared = width ** 2
		const totalSizeSquared = totalSize ** 2
```

File: visualization/app/codeCharta/util/algorithm/streetLayout/strip/verticalStrip.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	worstAspectRatio(nodes: CodeMapNode[], rect: Rectangle, rootSize: number, metricName: string): number {
+		const height = rect.height
+		const rootArea = rect.area()
+		const totalSize = this.totalScaledSize(nodes, metricName, rootSize, rootArea)
+		const stripMin = this.min(nodes, metricName, rootSize, rootArea)
+		const stripMax = this.max(nodes, metricName, rootSize, rootArea)
+		const heightSquared = Math.pow(height, 2)
+		const totalSizeSquared = Math.pow(totalSize, 2)
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/streetLayout/strip/verticalStrip.ts:54]:
```suggestion
		const heightSquared = height ** 2
		const totalSizeSquared = totalSize ** 2
```

File: visualization/app/codeCharta/util/algorithm/streetLayout/strip/verticalStrip.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+		return Math.max((heightSquared * stripMax) / totalSizeSquared, totalSizeSquared / (heightSquared * stripMin))
+	}
+}
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/streetLayout/strip/verticalStrip.ts:58]: These function seem very similar to some other functions. Maybe they could be abstracted a bit?

File: visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+				this.rightRow.push(child)
+			}
+		}
+	}
+
+	protected rearrangeRows(): void {
+		if (this.orientation === VerticalOrientation.UP) {
+			this.leftRow = [...this.leftRow].reverse()
+		} else {
+			this.rightRow = [...this.rightRow].reverse()
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts:143]:
```suggestion
			this.leftRow.reverse()
		} else {
			this.rightRow.reverse()
```

File: visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+				? streetBottomY - lastLeftNode.streetRect.getBottomRight().y
+				: this.height - this.getLength(this.leftRow)
+		const rightOverhang =
+			lastRightNode instanceof HorizontalStreet && lastRightNode.streetRect
+				? streetBottomY - lastRightNode.streetRect.getBottomRight().y
+				: this.height - this.getLength(this.rightRow)
+
+		return leftOverhang > 0 && rightOverhang > 0 ? Math.min(leftOverhang, rightOverhang) : 0
+	}
+}
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts:188]: The code has a lot of code that looks very similar. I did not yet check if it's possible to abstract it but we might want to have a closer look at that :)

File: visualization/app/codeCharta/util/algorithm/treeMapLayout/treeMapHelper.ts [hunk 9]
```diff
@@ -18,6 +18,17 @@ function countNodes(node: { children?: CodeMapNode[] }) {
...
 
+function calculateSize(node: CodeMapNode, metricName: string) {
+	let totalSize = node.attributes[metricName] || 0
+
+	if (totalSize === 0 && node.children && node.children.length > 0) {
+		for (const child of node.children) {
+			totalSize += calculateSize(child, metricName)
+		}
+	}
+	return totalSize
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/algorithm/treeMapLayout/treeMapHelper.ts:29]: This function looks identical to another one?

File: visualization/app/codeCharta/util/customConfigBuilder.ts [hunk 10]
```diff
@@ -121,7 +123,7 @@ function initializeTreeMapSettings(target: CustomConfig) {
 	}
 }
 
-function deepMapOneToOther<T>(source: any, target: T) {
+function deepMapOneToOther<T>(source: State, target: T) {
```
Comment (BridgeAR) [inline visualization/app/codeCharta/util/customConfigBuilder.ts:126]: ❤️

File: visualization/app/codeCharta/strretLayoutTodo.md [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+## bugs
+
+-    height is not calculated properly. 
+-    manual margin is not working. 
+
+## improvement
+
+-    scaling is not same in both algorithm.
+-    no label on streets.
+-    autocentering the camera does not always fit the label, the bounding box should include the labels.
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/strretLayoutTodo.md:15]: Yes, I added them

File: visualization/app/codeCharta/ui/codeMap/codeMap.render.service.ts [hunk 12]
```diff
@@ -39,7 +40,20 @@ export class CodeMapRenderService {
...
 	private getSortedNodes(map: CodeMapNode) {
 		const state = this.storeService.getState()
-		const nodes = createTreemapNodes(map, state, state.metricData.nodeMetricData, isDeltaState(state.files))
+		const {appSettings : {layoutAlgorithm}, metricData } = state
+		let nodes : Node[] = []
+		switch (layoutAlgorithm) {
+			case LayoutAlgorithm.StreetMap:
+				nodes = StreetLayoutGenerator.createStreetLayoutNodes(map, state, metricData.nodeMetricData, isDeltaState(state.files))
+				break
+			case LayoutAlgorithm.TMStreet:
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/ui/codeMap/codeMap.render.service.ts:49]: TMstreet is a street layout combined with the squarified layout

[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/ui/codeMap/codeMap.render.service.ts:49]: TMstreet is a street layout combined with the squarified layout

File: visualization/app/codeCharta/util/algorithm/streetLayout/horizontalStreet.ts [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	/**
+	 * Arranges rows according to their orientation
+	 */
+	protected rearrangeRows(): void {
+		if (this.orientation === HorizontalOrientation.RIGHT) {
+			this.bottomRow = [...this.bottomRow].reverse()
+		} else {
+			this.topRow = [...this.topRow].reverse()
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/horizontalStreet.ts:185]: Oops, my bad :)

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetLayoutGenerator.ts [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+	private static createStreet(node: CodeMapNode, orientation: StreetOrientation, children: BoundingBox[], depth: number) {
+		return orientation === StreetOrientation.Horizontal ?
+			new HorizontalStreet(node, children, depth) : new VerticalStreet(node, children, depth)
+	}
+ 
+	private static createTreeMap(node: CodeMapNode): TreeMap {
+		return new SquarifiedTreeMap(node)
+	}
+
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/streetLayoutGenerator.ts:83]: It definitely has to be removed !

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetLayoutGenerator.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+					children.push(street)
+				}
+			}
+		}
+		return children
+	}
+
+	private static createStreet(node: CodeMapNode, orientation: StreetOrientation, children: BoundingBox[], depth: number) {
+		return orientation === StreetOrientation.Horizontal ?
+			new HorizontalStreet(node, children, depth) : new VerticalStreet(node, children, depth)
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/streetLayoutGenerator.ts:77]: Can you please explain the benefit ?

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts (hunk 2 above)
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts:22]: Sure thing :)

File: visualization/app/codeCharta/util/algorithm/streetLayout/strip/horizontalStrip.ts (hunk 4 above)
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/strip/horizontalStrip.ts:54]: n1, didn't know about the **

File: visualization/app/codeCharta/codeCharta.model.ts [hunk 16]
```diff
@@ -45,6 +53,14 @@ export interface CodeMapNode {
 	fixedPosition?: FixedPosition
 }
 
+interface streetNode {
+	value?: number
+	rect?: Rectangle
+	zOffset?: number
+}
+export interface CodeMapNode extends squarifiedNode, streetNode{
+}
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/codeCharta.model.ts:62]: will check it

File: visualization/app/codeCharta/util/algorithm/streetLayout/strip/verticalStrip.ts (hunk 6 above)
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/strip/verticalStrip.ts:58]: I think there are a lot of refactoring that should be done, just wanted to push and we might refactor it later.

File: visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts (hunk 7 above)
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts:143]: my bad  :)

File: visualization/app/codeCharta/ui/maxTreeMapFiles/maxTreeMapFiles.component.html [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+			flex
+			step="1"
+			min="1"
+			max="1000"
+			type="number"
+			ng-model="$ctrl._viewModel.maxTreeMapFiles"
+			ng-change="$ctrl.onChangeMaxTreeMapFilesSlider()"
+		/>
+	</md-input-container>
+</md-slider-container>
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/ui/maxTreeMapFiles/maxTreeMapFiles.component.html:25]: This is for the TMstreet layout, as it combines the squarified layout to the streetview.

File: visualization/app/codeCharta/util/algorithm/streetLayout/boundingBox.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+	protected node: CodeMapNode
+	protected metricValue: number
+	protected FIXED_MARGIN = 0.5
+
+	constructor(node: CodeMapNode) {
+		this.node = node
+	}
+
+	public abstract calculateDimension(metricName: string): void
+	public abstract layout(origin: Vector2, margin: number): CodeMapNode[]
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/boundingBox.ts:17]: sur

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetLayoutGenerator.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (BridgeAR) [reply visualization/app/codeCharta/util/algorithm/streetLayout/streetLayoutGenerator.ts:77]: It's just a tad shorter code :) If you prefer the former version, that's also fine to me.

File: visualization/app/codeCharta/ui/fileChooser/fileChooser.e2e.ts [hunk 19]
```diff
@@ -52,7 +52,10 @@ describe("FileChooser", () => {
 	})
 
 	it("should open an valid and an invalid file, close the dialog and open a valid file", async () => {
-		await fileChooser.openFiles(["./app/codeCharta/assets/logo.png", "./app/codeCharta/assets/sample3.cc.json"])
+		// !note : openFiles is done in parallel @check Promise.all , first loaded is first served, 
+		// !an empty png file insure that the first loaded file is empty.png
+		// !fixes bug https://github.com/MaibornWolff/codecharta/issues/1322
+		await fileChooser.openFiles(["./app/codeCharta/assets/empty.png", "./app/codeCharta/assets/sample3.cc.json"])
```
[the remark this answers is not in the document]
Comment (Med-H) [reply visualization/app/codeCharta/ui/fileChooser/fileChooser.e2e.ts:58]: I'm also not sure why this happens, but I think it fixes the issue for now

File: visualization/app/codeCharta/util/algorithm/streetLayout/house.ts [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+
+export default class House extends BoundingBox {
+	constructor(node: CodeMapNode) {
+		super(node)
+	}
+
+	calculateDimension(metricName: string): void {
+		this.metricValue = StreetViewHelper.calculateSize(this.node, metricName)
+		this.width = Math.sqrt(this.metricValue)
+		this.height = Math.sqrt(this.metricValue)
```
[the remark this answers is not in the document]
Comment (ce-bo) [reply visualization/app/codeCharta/util/algorithm/streetLayout/house.ts:14]: I think you mean `const size`

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts (hunk 3 above)
Comment (Med-H) [reply visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts:28]: can't we get a child with a 0 metric value ? If not then I agree

Comment (BridgeAR) [reply visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts:28]: That's dealt with in my comment above :) https://github.com/MaibornWolff/codecharta/pull/1585#discussion_r542410449

File: visualization/app/codeCharta/ui/maxTreeMapFiles/maxTreeMapFiles.component.ts [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+		const maxTreeMapFiles = this.storeService.getState().appSettings.maxTreeMapFiles
+		this.onMaxTreeMapFilesChanged(maxTreeMapFiles)
+	}
+
+	onMaxTreeMapFilesChanged(maxTreeMapFiles: number) {
+		this._viewModel.maxTreeMapFiles = maxTreeMapFiles
+	}
+
+	onChangeMaxTreeMapFilesSlider() {
+		this.storeService.dispatch(setMaxTreeMapFiles(this._viewModel.maxTreeMapFiles))
```
Comment (ce-bo) [inline visualization/app/codeCharta/ui/maxTreeMapFiles/maxTreeMapFiles.component.ts:25]:
```suggestion
		// TODO: debounce dispatching the events
		this.storeService.dispatch(setMaxTreeMapFiles(this._viewModel.maxTreeMapFiles))
```

File: visualization/app/codeCharta/util/algorithm/streetLayout/horizontalStreet.ts [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+			child.calculateDimension(metricName)
+		}
+
+		// splits the children in top and bottom side of the street
+		this.splitChildrenToRows(this.children)
+
+		// note : longest street will be last in the row
+		this.rearrangeRows()
+
+		this.metricValue = StreetViewHelper.calculateSize(this.node, metricName)
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/algorithm/streetLayout/horizontalStreet.ts:46]:
```suggestion
		// TODO Add a comment what the following calculations are doing.
		this.metricValue = StreetViewHelper.calculateSize(this.node, metricName)
```

File: visualization/app/codeCharta/streetLayoutTodo.md [hunk 23]
```diff
@@ -0,0 +1,15 @@
+## refactoring
+
+-   refactoring of algorithm needs more work, maybe abstract the layout Generator for both street and squared layout.
+-   needs more unit tests.
```
Comment (ce-bo) [inline visualization/app/codeCharta/streetLayoutTodo.md:4]:
```suggestion
-   needs more unit tests.
-   refactore horizontalStreet.ts (layoutTopRow, layoutBottomRow are very similar)
```

File: visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts [hunk 24]
```diff
@@ -0,0 +1,93 @@
+import { Vector3 } from "three"
+import { CodeMapNode, Node, State } from "../../../codeCharta.model"
+import { getMarkingColor, isLeaf } from "../../codeMapHelper"
+import { getBuildingColor, getIncomingEdgePoint, isNodeFlat, isVisible, TreeMapHelper } from "../treeMapLayout/treeMapHelper"
+
+function calculateSize(node: CodeMapNode, metricName: string) {
+	// TODO if it is same as countNodes in treeMapHelper.ts
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/algorithm/streetLayout/streetViewHelper.ts:7]:
```suggestion
	// TODO if it is same as countNodes in treeMapHelper.ts
	// TODO from Ruben: This function is frequently used (even during sorting).
	//  As such, it's best to use memoization to reduce the computational overhead.
	//  That way there's no need to change the calling places of this function and it's still fast.
```

File: visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	calculateDimension(metricName: string): void {
+		for (const child of this.children) {
+			child.calculateDimension(metricName)
+		}
+		this.splitChildrenToRows(this.children)
+		this.rearrangeRows()
+
+		this.metricValue = StreetViewHelper.calculateSize(this.node, metricName)
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts:35]:
```suggestion
		// TODO refactor it - seems to be very similar to horizontalStreet.ts
		// TODO add a comment what the calculations are doing and why
		this.metricValue = StreetViewHelper.calculateSize(this.node, metricName)
```

File: visualization/app/codeCharta/util/algorithm/treeMapLayout/treeMapHelper.ts [hunk 26]
```diff
@@ -18,6 +18,17 @@ function countNodes(node: { children?: CodeMapNode[] }) {
 	return count
 }
 
+function calculateSize(node: CodeMapNode, metricName: string) {
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/algorithm/treeMapLayout/treeMapHelper.ts:21]:
```suggestion
// TODO this function exists twice in the code - please refactore it.
function calculateSize(node: CodeMapNode, metricName: string) {
```

**Comment by ce-bo** [comment]
PR merged. The TODOS should be fixed in a follow up Feature Request.

**Pull request merged**